Migrate Duration to smallest numeric types possible for efficiency

robot-head opened the pull request:
Addresses https://github.com/boa-dev/temporal/issues/189 by implementing the latest suggestion here: https://github.com/boa-dev/temporal/issues/189#issuecomment-2888385069

robot-head committed:
Use bnum to define an InternalDuration struct

robot-head committed:
Finish addressing all implications

robot-head committed:
Merge branch 'main' into ms/optimize_duration

robot-head committed:
Remove TimeDuration completely

robot-head committed:
Merge branch 'ms/optimize_duration' of https://github.com/robot-head/temporal into ms/optimize_duration

robot-head committed:
Fix compile error

robot-head committed:
Fix lint

robot-head committed:
Fix tests

robot-head committed:
Fix tests and lint errors

robot-head committed:
Fix depcheck

robot-head committed:
cargo run -p diplomat-gen

robot-head committed:
Fix sign calculations

robot-head committed:
Fix compile error

robot-head committed:
Fix duration sign handling and time diff

robot-head committed:
Merge pull request #6 from robot-head/codex/fix-failing-tests

Fix duration sign handling

robot-head committed:
Fix duration sign handling

robot-head committed:
Fix lint

robot-head committed:
Merge pull request #7 from robot-head/6o0p22-codex/fix-failing-tests

Fix duration sign handling

robot-head committed:
Fix lint

robot-head committed:
Fix duration calc

nekevss commented on src/iso.rs:
```diff
@@ -715,16 +724,30 @@ impl IsoTime {
-        TimeDuration::new_unchecked(h, m, s, ms, mis, ns)
+        let sign = duration_sign(&[h, m, s, ms, mis as i64, ns as i64]);
+
+        Duration::new_unchecked(
```
suggestion: return a NormalizedTimeDuration from `diff` instead of a duration.

I think that's a better adaption of the specification [here](https://tc39.es/proposal-temporal/#sec-temporal-differencetime)

mfa-ms replied:
Done

robot-head committed:
Respond to comments

mfa-ms commented on src/builtins/core/duration.rs:
```diff
@@ -66,15 +63,47 @@ impl PartialDuration {
     }
 }
 
+type U80 = BUintD16<5>; // 80 / 16 = 5
```
Done

mfa-ms commented on src/builtins/core/duration.rs:
```diff
@@ -101,47 +130,258 @@ impl core::fmt::Display for Duration {
-            TimeDuration::new_unchecked(value, 0, 0, 0, 0, 0),
-        )
+        Self {
+            sign: Sign::from(value),
```
Done

robot-head committed:
Merge remote-tracking branch 'origin/main' into ms/optimize_duration

robot-head committed:
Post-merge fixes

robot-head committed:
Fix sign on NormalizedDurationRecord

robot-head committed:
Fix nudge_to_day_or_time

Commit:
Add from_components method to NormalizedTimeDuration for creating instances from signed integer components

Commit:
Refactor U48 type definition to use BUintD16
Fix two unit tests try_from to use cast_from

Commit:
Refactor Duration constructor to use absolute values for time components

nekevss committed:
Merge branch 'main' into ms/optimize_duration

nekevss committed:
Fix broken tests + update various abstract operations

Many broken tests were related to DateDuration needing
to be an intermediate computational version. For instance,
days can exceed the normal threshold in some calculations.

There was also a number of operations that needed to be
updated the the current specification in order to fix the
bugs that came from this current update. Amongst them were
nudge_to_calendar_unit, nudge_to_day_or_time, various
InternalDuration record operations.

It's worth noting that this current update preserves
the old "normalized" specification language. Those should
probably be updated in follow ups to the newer language
for specification consistency, but that seemed to be out
of scope for the current PR.

nekevss committed:
Update the FFI

nekevss committed:
Update tzif-inspect + add test case to from_partial_duration

Manishearth commented on Cargo.toml:
```diff
@@ -82,6 +82,7 @@ timezone_provider = { workspace = true, optional = true }
 # System time feature
 web-time = { workspace = true, optional = true }
 iana-time-zone = { workspace = true, optional = true }
+bnum = "0.13.0"
```
issue: hmm, I know this was @nekevss's suggestion, but I'm not really excited about adding a new dependency just for an optimization, especially one we haven't measured the benefit of)

suggestion: Can we implement one of the earlier suggestions to use narrow stdlib integer types and flatten Duration, and then have a separate discussion for adding the `bnum` dependency? I imagine that switchover will be cleaner.


@nekevss opinions? 


(Given that this is reviewed already, I wouldn't recommend changing the PR until we decide what we want here, it's possible temporal_rs decides to take on the bnum dep)

nekevss replied:
Honestly, the one thing I've been most iffy about is adding the extra dependency (to the point of debating looking up how to manually implement arbitrarily sized integers 😅)

Hmmmmmm, I can flatten the struct and do some of the specification updates in a separate PR. If I realized that was going to be such a big portion of this PR, then I would've wanted that split off in a different PR anyways. My general concern is that the resulting Duration type is going to be huge.

Manishearth replied:
For a stack type, huge isn't really a big deal most of the time.

For JS implementors it will likely be heap allocated. But it's not like people are making thousands of these, I think.

Manishearth commented on temporal_capi/src/duration.rs:
```diff
@@ -56,42 +52,6 @@ pub mod ffi {
         }
     }
 
-    impl TimeDuration {
```
note: TimeDuration is not used in V8 right now, removing this API will not need complex migration

Manishearth commented on temporal_capi/src/instant.rs:
```diff
@@ -72,31 +72,30 @@ pub mod ffi {
-            duration: &TimeDuration,
-        ) -> Result<Box<Self>, TemporalError> {
+        // TODO: deprecate?
+        pub fn add_time_duration(&self, duration: &Duration) -> Result<Box<Self>, TemporalError> {
```
observation: this is fine to remove

Manishearth reviewed (commented):
I did not review the new code that appears to match the current spec in depth. I can if someone wants me to.

FFI changes look good, but I'm a bit worried about the additional dependency. I'm not sure it's pulling its weight.
